File: Cisco/8LongestNonDecreasingSubarrayFromTwoArrays.py
class Solution(object):
    def maxNonDecreasingLength(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: int
        """
        N = len(nums1)
        cur1, cur2 = 1, 1
        best = 1

        for i in range(1, N):  # O(N) time
            prev1, prev2 = cur1, cur2
            cur1, cur2 = 1, 1

            if nums1[i] >= nums1[i - 1]:
                cur1 = prev1 + 1
            if nums1[i] >= nums2[i - 1] and prev2 + 1 > cur1:
                cur1 = prev2 + 1

            if nums2[i] >= nums1[i - 1]:
                cur2 = prev1 + 1
            if nums2[i] >= nums2[i - 1] and prev2 + 1 > cur2:
                cur2 = prev2 + 1

            best = max(best, cur1, cur2)

        return best

File: Cisco/LinkedInList/C++/8LongestNonDecreasingSubarrayFromTwoArrays.cpp
#include <vector>

using namespace std;

class Solution
{
public:
    int maxNonDecreasingLength(vector<int> &nums1, vector<int> &nums2)
    {
        int N = nums1.size(), curr1, curr2, prev1 = 1, prev2 = 1, best = 1;

        for (int i = 1; i < N; ++i)
        {
            curr1 = 1, curr2 = 1;

            if (nums1[i] >= nums1[i - 1])
                curr1 = prev1 + 1;
            if (nums1[i] >= nums2[i - 1])
                curr1 = max(curr1, prev2 + 1);

            if (nums2[i] >= nums1[i - 1])
                curr2 = prev1 + 1;
            if (nums2[i] >= nums2[i - 1])
                curr2 = max(curr2, prev2 + 1);

            prev1 = curr1, prev2 = curr2, best = max(best, max(curr1, curr2));
        }

        return best;
    }
};